File: xmsinterp/interpolate/InterpLinearExtrapIdw.cpp
//------------------------------------------------------------------------------
/// \file
/// \brief Linear interpolation class with idw extrapolation
/// \copyright (C) Copyright Aquaveo 2020. Distributed under the xmsng
//------------------------------------------------------------------------------

//----- Included files ---------------------------------------------------------

// 1. Precompiled header

// 2. My header
#include <xmsinterp/interpolate/InterpLinearExtrapIdw.h>

// 3. Standard Library Headers

// 4. External Library Headers
#include <xmscore/misc/xmstype.h>

// 5. Shared Headers

// 6. Non-shared Headers

//----- Forward declarations ---------------------------------------------------

//----- External globals -------------------------------------------------------

//----- Namespace declaration --------------------------------------------------
namespace xms
{
//----- Constants / Enumerations -----------------------------------------------

//----- Classes / Structs ------------------------------------------------------


//----- Internal functions -----------------------------------------------------

//----- Class / Function definitions -------------------------------------------

////////////////////////////////////////////////////////////////////////////////
/// \class InterpLinearExtrapIdw
/// \brief Linear interpolation from a set of triangles
/// Uses the boost RTree for queries.
////////////////////////////////////////////////////////////////////////////////
//------------------------------------------------------------------------------
/// \brief
//------------------------------------------------------------------------------
InterpLinearExtrapIdw::InterpLinearExtrapIdw()
  : m_linear(InterpLinear::New())
  , m_idw(InterpIdw::New())
{
} // InterpLinearExtrapIdw::InterpLinearExtrapIdw
//------------------------------------------------------------------------------
/// \brief
//------------------------------------------------------------------------------
//InterpLinearExtrapIdw::~InterpLinearExtrapIdw()
//{
//} // InterpLinearExtrapIdw::~InterpLinearExtrapIdw
//------------------------------------------------------------------------------
/// \brief Creates an TriSearch class
/// \return Shared pointer to InterpLinear.
//------------------------------------------------------------------------------
BSHP<InterpLinearExtrapIdw> InterpLinearExtrapIdw::New()
{
  BSHP<InterpLinearExtrapIdw> ptr(new InterpLinearExtrapIdw());
  return ptr;
} // InterpLinear::New
//------------------------------------------------------------------------------
/// \brief Adds the triangles to the class
/// \param a_pts Array of point locations.
/// \param a_tris Array of triangles that references the a_pts array. This
/// array will have a size that is a multiple of 3. The first 3 locations in
/// array represent the first triangle and will have indices that correspond to
/// locations in the a_pts array.
//------------------------------------------------------------------------------
void InterpLinearExtrapIdw::SetPtsTris(BSHP<VecPt3d> a_pts, BSHP<VecInt> a_tris)
{
  m_linear->SetPtsTris(a_pts, a_tris);
  m_idw->SetPtsTris(a_pts, a_tris);
} // InterpLinearExtrapIdw::SetPtsTris
//------------------------------------------------------------------------------
/// \brief Set the scalars that will be used to interpolate from.
/// \param a_scalar Array of interpolation scalar values.
/// \param a_n Size of the array a_scalar.
//------------------------------------------------------------------------------
void InterpLinearExtrapIdw::SetScalars(const float* a_scalar, size_t a_n)
{
  m_linear->SetScalars(a_scalar, a_n);
  m_idw->SetScalars(a_scalar, a_n);
} // InterpLinearExtrapIdw::SetScalars
//------------------------------------------------------------------------------
/// \brief Set the scalars that will be used to interpolate from.
/// \param a_scalar: Array of interpolation scalar values.
//------------------------------------------------------------------------------
void InterpLinearExtrapIdw::SetScalars(BSHP<VecFlt> a_scalar)
{
  m_linear->SetScalars(a_scalar);
  m_idw->SetScalars(a_scalar);
} // InterpLinearExtrapIdw::SetScalars
//------------------------------------------------------------------------------
/// \brief Modifies the activity bitset of the class.
/// \param a_activity bitset of the activity of the points
//------------------------------------------------------------------------------
void InterpLinearExtrapIdw::SetPtActivity(DynBitset& a_activity)
{
  m_linear->SetPtActivity(a_activity);
  m_idw->SetPtActivity(a_activity);
} // InterpLinearExtrapIdw::SetPtActivity
//------------------------------------------------------------------------------
/// \brief Modifies the activity bitset of the class.
/// \param a_activity bitset of the activity of the points
//------------------------------------------------------------------------------
void InterpLinearExtrapIdw::SetTriActivity(DynBitset& a_activity)
{
  m_linear->SetTriActivity(a_activity);
} // InterpLinearExtrapIdw::SetTriActivity
//------------------------------------------------------------------------------
/// \brief Returns shared pointer to points vector.
/// \return Shared pointer to points vector.
//------------------------------------------------------------------------------
const BSHP<VecPt3d> InterpLinearExtrapIdw::GetPts() const
{
  return m_linear->GetPts();
} // InterpLinearExtrapIdw::GetPts
//------------------------------------------------------------------------------
/// \brief get point activity
/// \return bitset of point activity
//------------------------------------------------------------------------------
DynBitset InterpLinearExtrapIdw::GetPtActivity() const
{
  return m_linear->GetPtActivity();
} // InterpLinearExtrapIdw::GetPtActivity
//------------------------------------------------------------------------------
/// \brief 
/// \return 
//------------------------------------------------------------------------------
double InterpLinearExtrapIdw::GetExtrapVal() const
{
  return m_linear->GetExtrapVal();
} // InterpLinearExtrapIdw::GetExtrapVal
//------------------------------------------------------------------------------
/// \brief 
/// \return 
//------------------------------------------------------------------------------
bool InterpLinearExtrapIdw::GetTruncateInterpolatedValues() const
{
  return m_linear->GetTruncateInterpolatedValues();
} // InterpLinearExtrapIdw::GetTruncateInterpolatedValues
//------------------------------------------------------------------------------
/// \brief 
/// \return 
//------------------------------------------------------------------------------
double InterpLinearExtrapIdw::GetTruncMin() const
{
  return m_linear->GetTruncMin();
} // InterpLinearExtrapIdw::GetTruncMin
//------------------------------------------------------------------------------
/// \brief 
/// \return 
//------------------------------------------------------------------------------
double InterpLinearExtrapIdw::GetTruncMax() const
{
  return m_linear->GetTruncMax();
} // InterpLinearExtrapIdw::GetTruncMax
//------------------------------------------------------------------------------
/// \brief 
/// \return 
//------------------------------------------------------------------------------
bool InterpLinearExtrapIdw::GetUseCloughTocher() const
{
  return m_linear->GetUseCloughTocher();
} // InterpLinearExtrapIdw::GetUseCloughTocher
//------------------------------------------------------------------------------
/// \brief 
/// \return 
//------------------------------------------------------------------------------
bool InterpLinearExtrapIdw::GetUseNatNeigh() const
{
  return m_linear->GetUseNatNeigh();
} // InterpLinearExtrapIdw::GetUseNatNeigh
//------------------------------------------------------------------------------
/// \brief 
/// \return 
//------------------------------------------------------------------------------
int InterpLinearExtrapIdw::GetNatNeighNodalFunc() const
{
  return m_linear->GetNatNeighNodalFunc();
} // InterpLinearExtrapIdw::GetNatNeighNodalFunc
//------------------------------------------------------------------------------
/// \brief 
/// \return 
//------------------------------------------------------------------------------
int InterpLinearExtrapIdw::GetNatNeighNodalFuncNearestPtsOption() const
{
  return m_linear->GetNatNeighNodalFuncNearestPtsOption();
} // InterpLinearExtrapIdw::GetNatNeighNodalFuncNearestPtsOption
//------------------------------------------------------------------------------
/// \brief 
/// \return 
//------------------------------------------------------------------------------
int InterpLinearExtrapIdw::GetNatNeighNodalFuncNumNearestPts() const
{
  return m_linear->GetNatNeighNodalFuncNumNearestPts();
} // InterpLinearExtrapIdw::GetNatNeighNodalFuncNumNearestPts
//------------------------------------------------------------------------------
/// \brief 
/// \return 
//------------------------------------------------------------------------------
bool InterpLinearExtrapIdw::GetNatNeighBlendWeights() const
{
  return m_linear->GetNatNeighBlendWeights();
} // InterpLinearExtrapIdw::GetNatNeighBlendWeights
//------------------------------------------------------------------------------
/// \brief get triangle activity
/// \return bitset of triangle activity
//------------------------------------------------------------------------------
DynBitset InterpLinearExtrapIdw::GetTriActivity() const
{
  return m_linear->GetTriActivity();
} // InterpLinearExtrapIdw::GetTriActivity
//------------------------------------------------------------------------------
/// \brief Returns shared pointer to triangles vector.
/// \return Shared pointer to triangles vector.
//------------------------------------------------------------------------------
const BSHP<VecInt> InterpLinearExtrapIdw::GetTris() const
{
  return m_linear->GetTris();
} // InterpLinearExtrapIdw::GetTris
//------------------------------------------------------------------------------
/// \brief Returns shared pointer to scalars vector.
/// \return Shared pointer to scalars vector.
//------------------------------------------------------------------------------
const BSHP<VecFlt> InterpLinearExtrapIdw::GetScalars() const
{
  return m_linear->GetScalars();
} // InterpLinearExtrapIdw::GetScalars
//------------------------------------------------------------------------------
/// \brief Returns vector of point indexes for points that were outside of
/// all triangles
/// \return Returns vector of points indexes
//------------------------------------------------------------------------------
VecInt InterpLinearExtrapIdw::GetExtrapolationPointIndexes() const
{
  return m_linear->GetExtrapolationPointIndexes();
} // InterpLinearExtrapIdw::GetExtrapolationPointIndexes
//------------------------------------------------------------------------------
/// \brief Use the stored triangles to interpolate to a point. Returns
/// extrapolation value if the point is outside the triangles.
/// \param a_pt Location that is interpolated to.
/// \return Interpolated value at a_pt.
//------------------------------------------------------------------------------
float InterpLinearExtrapIdw::InterpToPt(const Pt3d& a_pt)
{
  float rval = XM_NODATA;
  if (m_linear->TriContainingPt(a_pt) > -1)
    rval = m_linear->InterpToPt(a_pt);
  else
    rval = m_idw->InterpToPt(a_pt);
  return rval;
} // InterpLinearExtrapIdw::InterpToPt
//------------------------------------------------------------------------------
/// \brief   Calls InterpToPt in a loop.
/// \param a_pts Locations of points.
/// \param a_scalars Interpolated scalar values at a_pts.
//------------------------------------------------------------------------------
void InterpLinearExtrapIdw::InterpToPts(const VecPt3d& a_pts, VecFlt& a_scalars)
{
  m_linear->InterpToPts(a_pts, a_scalars);
  VecInt extrap = m_linear->GetExtrapolationPointIndexes();
  for (size_t i = 0; i < extrap.size(); ++i)
  {
    a_scalars[extrap[i]] = m_idw->InterpToPt(a_pts[extrap[i]]);
  }
} // InterpLinearExtrapIdw::InterpToPts
//------------------------------------------------------------------------------
/// \brief Find the triangle containing the point
/// \param a_pt Location used to find a triangle.
/// \return Index of triangle containing a_pt. If XM_NONE is returned then no
/// triangle contained the point.
//------------------------------------------------------------------------------
int InterpLinearExtrapIdw::TriContainingPt(const Pt3d& a_pt)
{
  return m_linear->TriContainingPt(a_pt);
} // InterpLinearExtrapIdw::TriContainingPt
//------------------------------------------------------------------------------
/// \brief Find all triangle whose envelop contains the point.
/// \param a_pt Location used to find a triangle.
/// \param a_tris The indices to triangles whose envelop contains the point
//------------------------------------------------------------------------------
void InterpLinearExtrapIdw::TriEnvelopsContainingPt(const Pt3d& a_pt, VecInt& a_tris)
{
  m_linear->TriEnvelopsContainingPt(a_pt, a_tris);
} // InterpLinearExtrapIdw::TriContainingPt
//------------------------------------------------------------------------------
/// \brief Use the stored triangles to get interpolation weights for a point.
/// Returns false if the point is outside the triangles.
/// \param a_pt Location that is interpolated to.
/// \param a_idxs Vector of triangle point indices filled in by this method.
/// \param a_wts Vector of triangle point weights filled in by this method.
/// \return false if a_pt is not inside of any triangle
//------------------------------------------------------------------------------
bool InterpLinearExtrapIdw::InterpWeights(const Pt3d& a_pt, VecInt& a_idxs, VecDbl& a_wts)
{
  return m_linear->InterpWeights(a_pt, a_idxs, a_wts);
} // InterpLinearExtrapIdw::InterpWeights
//------------------------------------------------------------------------------
/// \brief Set the constant extrapolation value
/// \param a_val The value assigned to extrapolated points
//------------------------------------------------------------------------------
void InterpLinearExtrapIdw::SetExtrapVal(double a_val)
{
  m_linear->SetExtrapVal(a_val);
} // InterpLinearExtrapIdw::SetExtrapVal
//------------------------------------------------------------------------------
/// \brief Set the truncation values for the interpolation and turn on
/// truncation.
/// \param a_sMax The maximum value for truncation.
/// \param a_sMin The minimum value for truncation.
//------------------------------------------------------------------------------
void InterpLinearExtrapIdw::SetTrunc(double a_sMax, double a_sMin)
{
  m_linear->SetTrunc(a_sMax, a_sMin);
} // InterpLinearExtrapIdw::SetTrunc
//------------------------------------------------------------------------------
/// \brief Set the class to use the Clough Tocher interpolation method. This is
/// a legacy feature from GMS. Frankly it is so similar to linear that there is
/// really no reason to use it but it is here because users complain if features
/// are removed.
/// \param a_ True/False to indicate if CT should be used.
/// \param a_prog Progress bar to give users feed back on the set up process
/// of CT. If you have a really large set of triangles this may take some time.
//------------------------------------------------------------------------------
void InterpLinearExtrapIdw::SetUseCloughTocher(bool a_, BSHP<Observer> a_prog)
{
  m_linear->SetUseCloughTocher(a_, a_prog);
} // InterpLinearExtrapIdw::SetUseCloughTocher
//------------------------------------------------------------------------------
/// \brief Set the class to use natural neighbor (NN) interpolation.
/// \param a_ True/False to indicate if NN should be used.
/// \param a_nodalFunc Indicates which nodal function to use.
/// \param a_ndFuncOpt Indicates options for the nearest points when computing
/// the nodal functions.
/// \param a_ndFuncNumNearestPts The number of nearest points for nodal
/// function computation.
/// \param a_blendWeights Option to use a blending function on the calculated
/// weights.
/// \param a_prog Progress bar to give user feedback for generation of the
/// nodal functions.
//------------------------------------------------------------------------------
void InterpLinearExtrapIdw::SetUseNatNeigh(bool a_,
                                      int a_nodalFunc,
                                      int a_ndFuncOpt,
                                      int a_ndFuncNumNearestPts,
                                      bool a_blendWeights,
                                      BSHP<Observer> a_prog)
{
  m_linear->SetUseNatNeigh(a_, a_nodalFunc, a_ndFuncOpt, a_ndFuncNumNearestPts,
                           a_blendWeights, a_prog);
} // InterpLinearExtrapIdw::SetUseNatNeigh
//------------------------------------------------------------------------------
/// \brief Write the internals to a string.
/// \return The string.
//------------------------------------------------------------------------------
std::string InterpLinearExtrapIdw::ToString() const
{
  return m_linear->ToString();
} // InterpLinearExtrapIdw::ToString

} // namespace xms

#ifdef CXX_TEST
////////////////////////////////////////////////////////////////////////////////

#include <xmsinterp/interpolate/InterpLinearExtrapIdw.t.h>


// namespace xms {
using namespace xms;


////////////////////////////////////////////////////////////////////////////////
/// \class InterpLinearUnitTests
/// \brief tester for the InterpLinearUnitTests class
////////////////////////////////////////////////////////////////////////////////
//------------------------------------------------------------------------------
/// \brief tests doing interpolation to point outside the triangles
//------------------------------------------------------------------------------
void InterpLinearExtrapIdwUnitTests::testOutsideInterpolatePoint()
{
  BSHP<InterpLinear> interpolator = InterpLinearExtrapIdw::New();
  BSHP<VecPt3d> pts(new VecPt3d());
  *pts = {{0, 0, 0}, {10, 0, 1}, {10, 10, 2}, {0, 10, 3}};
  BSHP<VecInt> t(new VecInt());
  *t = {0, 1, 3, 1, 2, 3};

  interpolator->SetPtsTris(pts, t);

  Pt3d pt(0, 25, 0);

  float val = interpolator->InterpToPt(pt);
  TS_ASSERT_DELTA(2.70349288, val, FLT_EPSILON);
} // InterpLinearExtrapIdwUnitTests::testOutsideInterpolatePoint
//------------------------------------------------------------------------------
/// \brief tests doing interpolation to point inside the triangles
//------------------------------------------------------------------------------
void InterpLinearExtrapIdwUnitTests::testInsideInterpolatePoint()
{
  BSHP<InterpLinear> interpolator = InterpLinearExtrapIdw::New();
  BSHP<VecPt3d> pts(new VecPt3d());
  *pts = {{0, 0, 0}, {10, 0, 1}, {10, 10, 2}, {0, 10, 3}};
  BSHP<VecInt> t(new VecInt());
  *t = {0, 1, 3, 1, 2, 3};

  interpolator->SetPtsTris(pts, t);

  Pt3d pt(5, 5, 0);

  float val = interpolator->InterpToPt(pt);
  TS_ASSERT_DELTA(2.0, val, FLT_EPSILON);
} // InterpLinearExtrapIdwUnitTests::testInsideInterpolatePoint
//------------------------------------------------------------------------------
/// \brief tests doing interpolation to point inside the triangles
//------------------------------------------------------------------------------
void InterpLinearExtrapIdwUnitTests::testOneEdgeInterpolatePoint()
{
  BSHP<InterpLinear> interpolator = InterpLinearExtrapIdw::New();
  BSHP<VecPt3d> pts(new VecPt3d());
  *pts = {{0, 0, 0}, {10, 0, 1}, {10, 10, 2}, {0, 10, 3}};
  BSHP<VecInt> t(new VecInt());
  *t = {0, 1, 3, 1, 2, 3};

  interpolator->SetPtsTris(pts, t);

  Pt3d pt(10, 10, 0);

  float val = interpolator->InterpToPt(pt);
  TS_ASSERT_DELTA(2.0, val, FLT_EPSILON);
} // InterpLinearExtrapIdwUnitTests::testOneEdgeInterpolatePoint
//} // namespace xms
#endif // CXX_TEST

File: _package/xms/interp/interpolate/interp_linear_extrap_idw.py
"""Pure Python wrapping for InterpLinearExtrapIdw class."""
from .interpolator import Interpolator
from .._xmsinterp.interpolate import InterpLinearExtrapIdw as iLin


class InterpLinearExtrapIdw(Interpolator):
    """Class for performing linear interpolation."""
    nodal_function_types = {
        'constant': 0,
        'gradient_plane': 1,
        'quadratic': 2
    }

    nodal_function_point_search_options = {
        'natural_neighbor': 0,
        'nearest_points': 1,
    }

    def __init__(self, points=None, triangles=None, scalars=None, nodal_function=None,
                 point_search_option="natural_neighbor", number_nearest_points=16, blend_weights=True,
                 progress=None, quadrant_oct=False, weight_quadrant_oct=False,
                 weight_number_nearest_points=16, **kwargs):
        """Constructor.

        Args:
            points (list): The points of the source geometry as x,y,z tuples
            triangles (list): The triangles of the source geometry as point index tuples
            scalars (list): The scalar values at the points
            nodal_function (str): The nodal function to use One of: 'constant', 'gradient_plane', 'quadratic'
            point_search_option (str): The point search option. One of: 'natural_neighbor', 'nearest_points'
            number_nearest_points (int): The number of nearest points to consider
            blend_weights (bool): True if weights should be blended
            progress (Observer): Observer object for providing feedback
            weight_number_nearest_points (int): number of points to include in interpolation weight calculation
            quadrant_oct (bool): True if quadrant search should be used
            weight_quadrant_oct (bool): get nearest points from quadrants for weight calculation
            **kwargs (dict): Generic keyword arguments
        """
        if 'instance' in kwargs:
            self._instance = kwargs['instance']
            return

        if points is None:
            raise ValueError('"points" is a required argument.')
        self._check_points(points)

        if triangles is None or len(triangles) == 0:
            triangles = []
        else:
            self._check_triangles(triangles, len(points))

        if scalars is None or len(scalars) == 0:
            scalars = []
        else:
            self._check_scalars(scalars, len(points))

        self._instance = iLin(points, triangles, scalars)

        if nodal_function is not None:
            self.set_use_natural_neighbor(True, nodal_function, point_search_option, number_nearest_points,
                                          blend_weights, progress)
            self.set_idw_nodal_function(nodal_function, number_nearest_points, quadrant_oct, progress)
        self.set_idw_search_options(nearest_point=weight_number_nearest_points, quadrant_oct_search=weight_quadrant_oct)

        super().__init__(**kwargs)

    def __eq__(self, other):
        """Equality operator.

        Args:
            other (InterpLinearExtrapIdw): InterpLinearExtrapIdw to compare

        Returns:
            bool: True if InterpLinearExtrapIdws are equal
        """
        other_instance = getattr(other, '_instance', None)
        if not other_instance or not isinstance(other_instance, iLin):
            return False
        return other_instance == self._instance

    def __ne__(self, other):
        """Equality operator.

        Args:
            other (InterpLinearExtrapIdw): InterpLinearExtrapIdw to compare

        Returns:
            bool: True if InterpLinearExtrapIdws are not equal
        """
        result = self.__eq__(other)
        return not result

    def __repr__(self):
        """Returns a string representation of the InterpLinearExtrapIdw."""
        return '<InterpLinearExtrapIdw - Point Count: {}, Triangle Count: {}>'.format(
            len(self.points),
            int(len(self.triangles) / 3),
        )

    def __str__(self):
        """Returns a string representation of the InterpLinearExtrapIdw."""
        return '<InterpLinearExtrapIdw - Point Count: {}, Triangle Count: {}>'.format(
            len(self.points),
            int(len(self.triangles) / 3),
        )

    @staticmethod
    def _check_points(points):
        """Verifies that there are at least three points.

        Args:
            points (iterable): Array of points
        """
        if len(points) < 3:
            raise ValueError('"points" must be a list of 3 or more points')

    @staticmethod
    def _check_triangles(triangles, point_length):
        """Verifies that the triangles are valid.

        Args:
            triangles (iterable): Array of the triangles
            point_length (int): Count of all points
        """
        if len(triangles) % 3 != 0:
            raise ValueError('"triangles" must be a list of point indexes divisible by 3 defining each triangle.')
        max_triangles = max(triangles)
        if max_triangles > point_length - 1:
            raise ValueError('point in triangles is out of range: {}'.format(max_triangles))
        min_triangles = min(triangles)
        if min_triangles < 0:
            raise ValueError('point in triangles is out of range: {}'.format(min_triangles))

    @staticmethod
    def _check_scalars(scalars, point_length):
        """Verifies that the scalars are valid.

        Args:
            scalars (iterable): Array of scalars
            point_length (int): Count of all points
        """
        if len(scalars) != point_length:
            raise ValueError("Length of scalars must be equal to the length of points.")

    @staticmethod
    def _check_activity(activity, _length, _type):
        """Verifies that the activity is valid.

        Args:
            activity (iterable): Array of activity
            _length (int): Count of all points
            _type (string): 'points' or 'triangles'
        """
        if len(activity) != _length:
            raise ValueError("Length of activity must be equal to the length of {}.".format(_type))

    @staticmethod
    def _get_nodal_function_type(_str):
        """Gets the nodal function type based on the string.

        Args:
            _str (string): Nodal function type: 'constant', 'gradient_plane' or 'quadratic'

        Returns:
            The int value based on the nodal function type
        """
        nf_type = InterpLinearExtrapIdw.nodal_function_types.get(_str, None)
        if nf_type is None:
            raise ValueError('"nodal_function_type" must be one of {}, not {}'.format(
                ", ".join(InterpLinearExtrapIdw.nodal_function_types.keys()), _str))
        return nf_type

    @staticmethod
    def _get_nodal_function_point_search_options(_str):
        """Gets the nodal function point search option.

        Args:
            _str (string): Search option: 'natural_neighbors' or 'nearest_points'

        Returns:
            The int value based on the nodal function point search option
        """
        nf_type = InterpLinearExtrapIdw.nodal_function_point_search_options.get(_str, None)
        if nf_type is None:
            raise ValueError('"nodal_function_point_search_option" must be one of {}, not {}'.format(
                ", ".join(InterpLinearExtrapIdw.nodal_function_point_search_options.keys()), _str))
        return nf_type

    def set_points_and_triangles(self, points, triangles):
        """Adds the triangles to the class.

        Args:
            points (iterable): Array of point locations.
            triangles (iterable): Array of triangles that references the points array. This array will have a size that
                                  is a multiple of 3. The first 3 locations in array represent the first triangle and
                                  will have indices that correspond to locations in the points array.
        """
        self._instance.SetPtsTris(points, triangles)

    @property
    def points(self):
        """Gets the points."""
        return self._instance.GetPts

    @property
    def triangles(self):
        """Gets the triangles."""
        return self._instance.GetTris

    @property
    def scalars(self):
        """Gets the scalars."""
        return self._instance.GetScalars

    @scalars.setter
    def scalars(self, value):
        """Set the scalars that will be used to interpolate from.

        Args:
            value (iterable): Array of interpolation scalar values.
        """
        self._check_scalars(value, len(self.points))
        self._instance.SetScalars(value)

    def interpolate_to_point(self, point):
        """Use the stored triangles to interpolate to a point.

        Returns extrapolation value if the point is outside the triangles.

        Args:
            point (tuple): Location that is interpolated to.

        Returns:
            Interpolated value at point.
        """
        return self._instance.InterpToPt(point)

    def interpolate_to_points(self, points):
        """Calls InterpToPt in a loop.

        Args:
            points (iterable): Locations of points.

        Returns:
            Interpolated values at points.
        """
        return self._instance.InterpToPts(points)

    @property
    def point_activity(self):
        """Gets the point activity."""
        return self._instance.GetPtActivity

    @point_activity.setter
    def point_activity(self, value):
        """Modifies the activity bitset of the class.

        Args:
            value (iterable): Bitset of the activity of the points
        """
        self._check_activity(value, len(self.points), "points")
        self._instance.SetPtActivity(value)

    @property
    def triangle_activity(self):
        """Gets the triangle activity."""
        return self._instance.GetTriActivity

    @triangle_activity.setter
    def triangle_activity(self, value):
        """Modifies the activity bitset of the class.

        Args:
            value (iterable): Bitset of the activity of the triangles
        """
        self._check_activity(value, len(self.triangles) / 3, "triangles")
        self._instance.SetTriActivity(value)

    @property
    def extrapolation_point_indexes(self):
        """Returns vector of point indexes for points that were outside of all triangles."""
        return self._instance.GetExtrapolationPointIndexes

    def triangle_containing_point(self, point):
        """Find the triangle containing the point.

        Args:
            point (tuple): Location used to find a triangle.

        Returns:
            Index of triangle containing point. If None is returned then no triangle contained the point.
        """
        return self._instance.TriContainingPt(point)

    def triangle_envelopes_containing_point(self, point):
        """Find all triangles whose envelop contains the point.

        Args:
            point (tuple): Location used to find a triangle.

        Returns:
            Iterable of triangle indexes containing the point
        """
        return self._instance.TriEnvelopsContainingPt(point)

    def interpolate_weights(self, point):
        """Use the stored triangles to get interpolation weights for a point.

        Returns False if the point is outside the triangles.

        Args:
            point (tuple): Location that is interpolated to.

        Returns:
            tuple of a bool, array of indexes, and array of weights
        """
        return self._instance.InterpWeights(point)

    @property
    def extrapolation_value(self):
        """Set the constant extrapolation value."""
        return self._instance.GetExtrapVal

    @extrapolation_value.setter
    def extrapolation_value(self, value):
        """Get extrapolation value."""
        self._instance.SetExtrapVal(value)

    def set_truncation(self, maximum, minimum):
        """Set the truncation values for the interpolation and turn on truncation.

        Args:
            maximum (float): The maximum value for truncation.
            minimum (float): The minimum value for truncation.
        """
        if maximum < minimum:
            raise ValueError('maximum must be greater than minimum')
        self._instance.SetTrunc(maximum, minimum)

    @property
    def use_clough_tocher(self):
        """Get the option for using Clough Tocher interpolation."""
        return self._instance.GetUseCloughTocher()

    def set_use_clough_tocher(self, on, progress=None):
        """Set the class to use the Clough Tocher interpolation method.

        This is a legacy feature from GMS. Compare to linear.

        Args:
            on (bool): TIndicate if CT should be used.
            progress (observer): Progress bar to give users feed back on the set up process of CT.
                                 If you have a really large set of triangles this may take some time.
        """
        self._instance.SetUseCloughTocher(on, progress)

    @property
    def use_natural_neighbor(self):
        """Get the option for using Natural Neighbor interpolation."""
        return self._instance.GetUseNatNeigh()

    def set_use_natural_neighbor(self, on=True, nodal_function="constant",
                                 point_search_option="natural_neighbor",
                                 number_nearest_points=16, blend_weights=True,
                                 progress=None):
        """Set the class to use natural neighbor (NN) interpolation.

        Args:
            on (bool): Indicate if NN should be used.
            nodal_function (str): Indicates which nodal function to use.
            point_search_option (str): Indicates options for the nearest points when computing
                                                      the nodal functions.
            number_nearest_points (int): The number of nearest points for nodal function computation.
            blend_weights (bool): Option to use a blending function on the calculated weights.
            progress (observer): Progress bar to give user feedback for generation of the nodal functions.
        """
        nft = self._get_nodal_function_type(nodal_function)
        nfpso = self._get_nodal_function_point_search_options(point_search_option)
        self._instance.SetUseNatNeigh(on, nft, nfpso, number_nearest_points,
                                      blend_weights, progress)

    def set_idw_nodal_function(self, nodal_function="constant", number_nearest_points=16, quadrant_oct=False,
                               progress=None):
        """Sets the type of nodal function as well as options for computing nodal functions.

        Args:
            nodal_function (string): The nodal function methodology:
                                          constant (0), gradient plane (1), quadratic (2).
            number_nearest_points (int): The nearest number of points to use when calculating the nodal functions.
            quadrant_oct (bool): Find the nearest number of points in each quadrant (2d) or octant (3d) when computing
                                 nodal functions.
            progress (float): Progress bar to give user feedback.
        """
        nft = self._get_nodal_function_type(nodal_function)
        self._instance.SetIdwNodalFunction(nft, number_nearest_points, quadrant_oct, progress)

    def set_idw_search_options(self, nearest_point, quadrant_oct_search):
        """Sets the search options for how to find the nearest points to the interpolation point.

        The number of nearest points can be specified as well as whether to find the nearest points in each
        quadrant or octant.

        Args:
            nearest_point (int): The number of nearest points to the interpolation point. These points are used to do
                                 the interpolation.
            quadrant_oct_search (bool): Specifies if the search criterion should find the nearest points in each
                                 quadrant (2d) or octant (3d)
        """
        self._instance.SetIdwSearchOpts(nearest_point, quadrant_oct_search)

    @property
    def truncate_interpolated_values(self):
        """Get the option to truncate interpolated values."""
        return self._instance.GetTruncateInterpolatedValues

    @property
    def truncate_min(self):
        """Get minimum truncation value."""
        return self._instance.GetTruncMin

    @property
    def truncate_max(self):
        """Get maximum truncation value."""
        return self._instance.GetTruncMax

    @property
    def native_neighbor_nodal_function(self):
        """Get the value for the Natural Neighbor nodal function."""
        return self._instance.GetNatNeighNodalFunc

    @property
    def native_neighbor_nodal_function_nearest_points_option(self):
        """Get the option for the Natural Neighbor nodal function nearest points."""
        return self._instance.GetNatNeighNodalFuncNearestPtsOption

    @property
    def native_neighbor_nodal_function_number_nearest_points(self):
        """Get the value for the number of nearest points to use when calculating the nodal function."""
        return self._instance.GetNatNeighNodalFuncNumNearestPts

    @property
    def native_neighbor_blend_weights(self):
        """Get the option for blending weights when using Natural Neighbor."""
        return self._instance.GetNatNeighBlendWeights
